### PR TITLE
RSDK-7821 Log basic metrics from signaling answerer

### PR DESCRIPTION
**Pull request by benjirewis**
RSDK-7821

Logs information about each connection establishment attempt from the signaling answerer through structure-logging a new `answererStats` struct with information filled in throughout the course of the `answer()` method.

### CODE REVIEW

File: rpc/wrtc_signaling_answerer.go
Comment (benjirewis) [inline rpc/wrtc_signaling_answerer.go:259]: This log looks like:
```
2024-04-24T14:17:49.058-0400	INFO	robot_server.external_signaler	rpc/wrtc_signaling_answerer.go:259	Connection establishment success	{"stats": "{\"answer_request_init_received\":\"2024-04-24T14:17:48.80109-04:00\",\"num_answer_updates\":3,\"average_answer_update\":29236,\"max_answer_update\":49500,\"local_ice_candidates\":[{\"gathered_at\":\"2024-04-24T14:17:48.803465-04:00\",\"candidate\":\"udp4 host 127.0.0.1:57413\"},{\"gathered_at\":\"2024-04-24T14:17:48.803721-04:00\",\"candidate\":\"udp4 host 192.168.1.127:50619\"},{\"gathered_at\":\"2024-04-24T14:17:48.861006-04:00\",\"candidate\":\"udp4 srflx 66.24.146.163:61077 related 0.0.0.0:61077\"}],\"remote_ice_candidates\":[{\"received_at\":\"2024-04-24T14:17:48.987992-04:00\",\"candidate\":\"candidate:392662914 1 udp 2130706431 859ad36c-1539-485d-bb53-046e3321f263.local 63388 typ host\"},{\"received_at\":\"2024-04-24T14:17:48.995173-04:00\",\"candidate\":\"candidate:1938624796 1 udp 1694498815 66.24.146.163 57236 typ srflx raddr 0.0.0.0 rport 57236\"},{\"received_at\":\"2024-04-24T14:17:48.997028-04:00\",\"candidate\":\"candidate:1938624796 1 udp 1694498815 66.24.146.163 58063 typ srflx raddr 0.0.0.0 rport 58063\"},{\"received_at\":\"2024-04-24T14:17:48.997058-04:00\",\"candidate\":\"candidate:2838388448 1 udp 16777215 34.203.251.220 28383 typ relay raddr 0.0.0.0 rport 64553\"},{\"received_at\":\"2024-04-24T14:17:48.997071-04:00\",\"candidate\":\"candidate:392662914 1 udp 2130706431 859ad36c-1539-485d-bb53-046e3321f263.local 59069 typ host\"}]}"}
```

**Review by benjirewis — COMMENTED** [review]
This is still a bit of a WIP, but I thought I'd pause for some review. Will leave in "draft" for the time being.

File: rpc/wrtc_signaling_answerer.go
Comment (cheukt) [reply rpc/wrtc_signaling_answerer.go:259]: can we make it so the fields of the blob is spread (so like `ans.logger.Infow("Connection establishment success", "answer_request_init_received", xxxx, "num_answer_updates", x)

tho I don't feel that strongly about it. I think it might be slightly easier to read and parse for upstream, but not by much.

**Comment by cheukt** [comment]
will defer to you in terms of what is most useful to collect, but the way of collecting seems fine

File: rpc/wrtc_signaling_answerer.go
Comment (benjirewis) [reply rpc/wrtc_signaling_answerer.go:259]: Yeah it's interesting; logging the stringified JSON blob here is pretty terrible in a readability sense. But, if we start to use Go structured logging with `Infow`, there are a couple issues:
- https://viam.atlassian.net/browse/APP-4143 means many of these fields won't show up in the `Logs` tab and we'll have to look at the `logs` DB for the time being to see the real values
- There will be no real Go "structure" representing the stats, only key-value pairs that are potentially more subject to change

So, I'm torn. Neither solution is really optimal IMO, but maybe this JSON blob is acceptable for the time being until we potentially start sending these metrics to some metric collector and stop logging them here? To be clear, a log similar to the one I've copied above will show up in RDK logs every time the robot succeeds or fails a connection establishment with a client. Do we think users will find that troubling/annoying?

File: rpc/wrtc_signaling_answerer.go
Comment (benjirewis) [inline rpc/wrtc_signaling_answerer.go:252]: For some reason one answerer is outputting `Connection establishment succeeded` and the other is outputting `Connection establishment failed` both with similar stats for a single connection establishment. Looking into that before I pull this out of "draft"; makes me think I'm misunderstanding how the two answerers interact.

**Review by benjirewis — COMMENTED** [review]
Switched to Go structured logging (still hypothetically a single `INFO`-level message for each conn. establishment attempt) instead of JSON as discussed offline.

File: rpc/wrtc_signaling_answerer.go
Comment (benjirewis) [reply rpc/wrtc_signaling_answerer.go:252]: I believe this was due to RSDK-7311 and should be fixed by #283 from my testing. `app`'s goutils would have to have #283 in order and `rdk` would have to have this PR's changes to have a single `Connection establishment [succeeded|failed]` message from a running RDK. Curious if we want to wait for `app` to upgrade its `goutils` version to one that contains #283 (if reviewers agree on merging that PR) before merging this PR/upgrading the `goutils` version in `rdk`, @cheukt.

Comment (cheukt) [reply rpc/wrtc_signaling_answerer.go:252]: would wait until #283 is mergedbefore upgrading goutils in rdk

Comment (benjirewis) [reply rpc/wrtc_signaling_answerer.go:252]: Will move this PR out of draft but won't merge until we fix the double `AnswerRequest_Init` issue.

Comment (benjirewis) [inline rpc/wrtc_signaling_answerer.go:252]: This log now looks like:
```
2024-04-29T14:03:46.512-0400	INFO	robot_server.external_signaler	rpc/answerer_stats.go:72	Connection establishment succeeded	{"answerRequestInitReceieved": "2024-04-29T14:03:46.280-0400", "numAnswerUpdates": 4, "averageAnswerUpdate": "172.614µs", "maxAnswerUpdate": "431.875µs", "localICECandidates": ["at: 2024-04-29T14:03:46.287-0400, candidate: udp4 host 127.0.0.1:62814", "at: 2024-04-29T14:03:46.287-0400, candidate: udp4 host 192.168.10.28:58479", "at: 2024-04-29T14:03:46.287-0400, candidate: udp4 host 192.168.64.1:55785", "at: 2024-04-29T14:03:46.343-0400, candidate: udp4 srflx 24.61.184.231:59442 related 0.0.0.0:59442"], "remoteICECandidates": ["at: 2024-04-29T14:03:46.449-0400, candidate: candidate:2894675966 1 udp 1694498815 24.61.184.231 55170 typ srflx raddr 0.0.0.0 rport 55170", "at: 2024-04-29T14:03:46.449-0400, candidate: candidate:2894675966 1 udp 1694498815 24.61.184.231 52601 typ srflx raddr 0.0.0.0 rport 52601", "at: 2024-04-29T14:03:46.449-0400, candidate: candidate:1308079556 1 udp 2130706431 6e5bce95-ff44-4118-a1a5-34ea9384146b.local 50546 typ host", "at: 2024-04-29T14:03:46.449-0400, candidate: candidate:1308079556 1 udp 2130706431 6e5bce95-ff44-4118-a1a5-34ea9384146b.local 52462 typ host", "at: 2024-04-29T14:03:46.449-0400, candidate: candidate:1196105838 1 udp 16777215 34.203.251.176 27900 typ relay raddr 0.0.0.0 rport 62505", "at: 2024-04-29T14:03:46.456-0400, candidate: candidate:1308079556 1 udp 2130706431 6e5bce95-ff44-4118-a1a5-34ea9384146b.local 61301 typ host"]}
```

Those measurements of answer update timings are very short; that's because sending an `AnswerResponse_Update` back to the signaling server does not include the time taken to actually update the database; we're using a bidi stream here, so the sending of a message does not include waiting for any response from the other side.

Comment (cheukt) [reply rpc/wrtc_signaling_answerer.go:252]: would it make sense to try/is it possible to measure the time to update the database? related question is, how useful is the number if it only captures the time to send the request and nothing after?

Comment (benjirewis) [reply rpc/wrtc_signaling_answerer.go:252]: > to measure the time to update the database

From the answerer side, I don't think this is possible given the current setup. I think the timing we care about is a combination of timings from the answerer and the client:
- When did the answerer generate the candidate?
- When did the answerer send said candidate to the signaling server?
- When did the client receive said candidate from the signaling server?

Those would be understood by looking at the timestamps of an SDK and the RDK it's connecting to.

> how useful is the number if it only captures the time to send the request and nothing after?

Yeah good question: I think not that useful. I'll remove them for now.

**Comment by cheukt** [comment]
a bit of a nit, can you upload the log with the JSON formatted? so each field is a newline

**Comment by benjirewis** [comment]
> a bit of a nit, can you upload the log with the JSON formatted? so each field is a newline

Yep:

```
2024-04-29T17:47:14.861-0400	INFO	robot_server.external_signaler	rpc/answerer_stats.go:67	Connection establishment succeeded	

{
  "answerRequestInitReceieved": "2024-04-29T17:47:14.643-0400",
  "numAnswerUpdates": 3,
  "localICECandidates": [
    "at: 2024-04-29T17:47:14.647-0400, candidate: udp4 host 127.0.0.1:64436",
    "at: 2024-04-29T17:47:14.647-0400, candidate: udp4 host 192.168.10.28:52064",
    "at: 2024-04-29T17:47:14.681-0400, candidate: udp4 srflx 24.61.184.231:53612 related 0.0.0.0:53612"
  ],
  "remoteICECandidates": [
    "at: 2024-04-29T17:47:14.795-0400, candidate: candidate:1115398635 1 udp 2130706431 5e05cd36-52a1-4853-a0b2-e8deb981b3b6.local 59730 typ host",
    "at: 2024-04-29T17:47:14.795-0400, candidate: candidate:2894675966 1 udp 1694498815 24.61.184.231 63504 typ srflx raddr 0.0.0.0 rport 63504",
    "at: 2024-04-29T17:47:14.817-0400, candidate: candidate:3792049949 1 udp 16777215 34.203.251.122 18949 typ relay raddr 0.0.0.0 rport 54933",
    "at: 2024-04-29T17:47:14.817-0400, candidate: candidate:1115398635 1 udp 2130706431 5e05cd36-52a1-4853-a0b2-e8deb981b3b6.local 62737 typ host",
    "at: 2024-04-29T17:47:14.828-0400, candidate: candidate:2894675966 1 udp 1694498815 24.61.184.231 50752 typ srflx raddr 0.0.0.0 rport 50752"
  ]
}
```

File: rpc/answerer_stats.go
Comment (cheukt) [inline rpc/answerer_stats.go:30]: local refers to client side and remote would be server side? or is it the other way around? might be good to clarify here

**Review by cheukt — APPROVED** [review]
looks great! excited to see what we gather

File: rpc/answerer_stats.go
Comment (benjirewis) [reply rpc/answerer_stats.go:30]: Ah yeah confusing. Local is the signaling answerer's "local", so technically server-side. Remote is what's received from the client, so client-side. Will add a comment to clarify.

File: rpc/wrtc_signaling_answerer.go
Comment (benjirewis) [reply rpc/wrtc_signaling_answerer.go:252]: Confirmed since `app` has been upgraded to have #283, the double `AnswerRequest_Init` issue is gone.

**Review by ale7714 — APPROVED** [review]
Looks great! Thank you for the comments on AnswererStats super helpful 🚀

**Pull request merged**
